crates/ruff_linter/src/rules/pycodestyle/rules/too_many_newlines_at_end_of_file.rs

```diff
@@ -0,0 +1,87 @@
+use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_parser::lexer::LexResult;
+use ruff_python_parser::Tok;
+use ruff_text_size::{TextRange, TextSize};
+
+/// ## What it does
+/// Checks for files with multiple trailing blank lines.
+///
+/// ## Why is this bad?
+/// Trailing blank lines in a file are superfluous.
+///
+/// However, the last line of the file should end with a newline.
+///
+/// ## Example
+/// ```python
+/// spam(1)\n\n\n
+/// ```
+///
+/// Use instead:
+/// ```python
+/// spam(1)\n
+/// ```
+#[violation]
+pub struct TooManyNewlinesAtEndOfFile {
+    /// The number of trailing newlines.
+    count: u32,
+}
+
+impl AlwaysFixableViolation for TooManyNewlinesAtEndOfFile {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let TooManyNewlinesAtEndOfFile { count } = self;
+        if *count > 1 {
+            format!("Too many newlines at end of file")
+        } else {
+            format!("Extra newline at end of file")
+        }
+    }
+
+    fn fix_title(&self) -> String {
+        let TooManyNewlinesAtEndOfFile { count } = self;
+        if *count > 1 {
+            "Remove trailing newlines".to_string()
+        } else {
+            "Remove trailing newline".to_string()
+        }
+    }
+}
+
+/// W391
+pub(crate) fn too_many_newlines_at_end_of_file(
+    diagnostics: &mut Vec<Diagnostic>,
+    lxr: &[LexResult],
```

Comment:
I removed the `Locator`. I don't think we need to guard on empty files here -- that's just for the "no trailing newline" check, since empty files would be a false positive.